### ISSUE TITLE
Python 3.11

### ISSUE THREAD

**Issue by esc** [numba/numba#8304]
This is the meta/coordination issue for Python 3.11 support for Numba and llvmlite.

Note: The issue text will be updated continuously to reflect the current process and progress. 

As Numba interfaces with cpython on the bytecode and C-level, Python minor version upgrades present a significant compatibility challenge and have -- in the past -- required multiple PRs over multiple months of work. For Python 3.11 this issue will serve as a central point of coordination of the efforts. As such it will list the current branches pertaining to Python 3.11 support in addition to using the tag `Python 3.11`.

In order to support testing against beta releases and release candidates, I created a small utility that allows easy setup of a Numba development environment against a Python pre-release docker container. https://github.com/numba/numba-hatchery this can be used with development branches/repositories of Numba and llvmlite.

* llvmlite status
  * Using the PR https://github.com/numba/llvmlite/pull/869 passes all unit tests when used with numba-hatchery

* numba status
  * Numba Python 3.11 support is now being worked on across multiple branches:
    * https://github.com/esc/numba/tree/python3.11_partial/version_bump  -- version bump
    * https://github.com/esc/numba/tree/python3.11_partial/compile-hack -- compile hacks (WIP)
    * https://github.com/esc/numba/tree/python3.11_partial/implement_resume -- RESUME implementation
  * Numba would not compile initially. The https://github.com/esc/numba/tree/python3.11_partial/compile-hack branch  was created to allow. Numba to compile
  * The new `RESUME` opcode wasn't being handled. https://github.com/esc/numba/tree/python3.11_partial/implement_resume  implements this.
  * The branch `esc/python3.11` now contains a merge of all involved feature branches and can be used for testing all required patches in unison: https://github.com/numba/numba/pull/8381
  * @sklam is making progress with the bytecode at: https://github.com/sklam/numba/tree/py3.11wip

* TODO
  * Cleanup the compile-hacks PR so that it is sane and not just something to allow compilation
  * Port distutils to use setuptools as per https://peps.python.org/pep-0632/

**Comment by sklam** [comment]
I'm making progress on the bytecode support at https://github.com/sklam/numba/tree/py3.11wip

**Comment by sklam** [comment]
I'm making progress at https://github.com/numba/numba/pull/8545. The PR adapts Numba for the major bytecode changes in Py3.11.

**Comment by binary-husky** [comment]
Still no official support for python 3.11 ?

**Comment by alonme** [comment]
Any news?

**Comment by extesy** [comment]
There is an active https://github.com/numba/numba/commits/py3.11 branch, a lot of progress is happening.

**Comment by esc** [comment]
@extesy thank you for listing that link. @alonme to answer your question: Indeed all the work for Python 3.11 is currently being merged one PR at a time to the `py3.11` branch and collected there for further testing. IIRC @sklam mentioned during the last developer meeting (https://github.com/numba/numba/wiki/Minutes_2022_11_29) that there are only a handful of failing tests that remain. From what I can tell it shouldn't be long now and with any luck there will be a Numba release candidate that supports 3.11 before the end of the year 🤞 (but making no guarantees, obviously).

**Comment by alonme** [comment]
@esc thanks for the detailed answer!
Waiting for it

**Comment by stuartarchibald** [comment]
The changes needed to accommodate the majority of the new Python 3.11 op-code additions, along with changes to the op-code behaviour/sequences in the bytecode, have been addressed in #8545 and #8590 with thanks to @sklam. The majority of the remaining issues are addressed in #8639, a summary of the current state of the Numba test suite against this branch is here https://github.com/numba/numba/pull/8639#issuecomment-1337291739 (less than 10 failures with varying degrees of difficulty to fix).

**Comment by jschwar313** [comment]
Is there a pip command available to install numba on python 3.11.1?

**Comment by jamesmyatt** [comment]
> Is there a pip command available to install a dev version of numba on python 3.11.1?

Look at https://pip.pypa.io/en/stable/topics/vcs-support/ and Example #5 here https://pip.pypa.io/en/stable/cli/pip_install/#examples

**Comment by jschwar313** [comment]
Yea, that part I know.  What's the github location of the software I would install that would replace the numba package?  That I don't know.  Thanks.

**Comment by sklam** [comment]
If you want to pip install via the vcs support, you can do so via `pip install git+https://github.com/numba/numba@py3.11`.  But bewared that installing llvmlite is not as easy due to llvm. That is why we provided a docker setup for testing. The easiest way to get the llvm is still by using our `llvmdev` conda package.

**Comment by MarcusTherkildsen** [comment]
6 months and still no sign of a release. What's the hold up?

**Comment by esc** [comment]
> 6 months and still no sign of a release. What's the hold up?

Thank you for asking about this. To clarify, the hold-up comes from two sources:

* Waiting on anaconda-official Python 3.11 conda packages, inkl. dependencies (Numpy, etc..) -- Python 3.11 is available at the time of writing, but the dependencies are missing. Our build and release pipelines depend on anaconda-conda packages for the time being and while there are conda-forge packages, there are too unstable for our needs and have caused more issues than solved ones in the past. Upgrading the pipeline for some other build system is not an option due to resource constraints.

* The need to update the Numba code-base to support Python 3.11. Numba operates on Python bytecode and this is essentially a moving target. It takes between 3-6 months of work for every minor Python release to make Numba compatible with the new bytecode version. Bytecode isn't designed as an interface, so both the number of bytecodes and how each one behaves changes significantly with every Python minor Version. For Python 3.11 we are almost there and there are only a handful of tests still failing, last time I checked.

In the name of the Numba developers I would like to re-iterate your continued patience on this front. It's not long to go now and based on today's evidence I expect a Numba release candidate to become available in 3-6 weeks. 🙏

**Comment by MarcusTherkildsen** [comment]
That sounds great, thanks for the update!

**Comment by esc** [comment]
Just a quick update here. We are still working on this. The Python 3.11 support is now complete and we are waiting on certain Numba dependencies to be packaged as conda packages via anaconda.org (can be tracked at: https://github.com/ContinuumIO/anaconda-issues/issues/13082). Most things for the next release are in place and things are moving forward. Thank you again everyone for your sustained patience!!

**Comment by thebaptiste** [comment]
May be you should consider using something else than conda for your CI ? 
Your code seems to be ready, your dependencies are also ready (tarballs or wheels), but you have to wait for conda packages, what a waste of time !
May be you could considerer providing a tarball pre-release ?

**Comment by tanlin2013** [comment]
Perhaps we could also release on pip first. That will be appreciated too!

**Comment by seibert** [comment]
Just to clarify: This is definitely something we want to improve, but the challenge here is in testing and validating that Numba works in every way it needs to in every environment we support.  Putting Numba into a package file (conda or wheel or whatever) is relatively more straightforward, but that doesn't ensure it will work.

We have relied on conda mostly because it enabled us to easily create 100+ different environment configurations across different combinations of CPU platform (x86, arm, ppc), operating system (win, mac, linux), numpy (several here), as well as to check parallel execution with various threading backends (OpenMP, TBB, etc), along with different CUDA versions.  One we are sure all that is OK, then making a final wheel/sdist is not much trouble.

So the take away is that we definitely hear the feedback, but are seeking a way to ensure we can efficiently test Numba so that when a new release is posted, it will work for everyone.  We have some ideas and will try some process changes this year.

**Comment by MarcusTherkildsen** [comment]
Hi again

25th of January you said:

"It's not long to go now and based on today's evidence I expect a Numba release candidate that supports Python 3.11 to become available in 3-6 weeks. 🙏"

We are closing on 6 weeks now (in two days). What is the status? 
Thanks for your hard work, it makes our work a lot easier :)

**Comment by esc** [comment]
@MarcusTherkildsen thank you for asking about this. I asked around today and here is what I gather: 

* the packages for 3.11 should arrive early next week as @cbouss will be able to acknowledge
* the dev branches for the support of Numpy 1.24, LLVM 14 and Python 3.11 are ready to merge
* @sklam has built some internal testing around conda-forge packages that will allow us to perform tests and attempt the aforementioned merges

So the TL;DR: progress has been made and chances are good to get a release candidate next week or the week after, barring any more major issues arising during integration of the dev branches. 

Also, It looks like @sklam will be the release manager for this one -- so please feel free to add more context in case I forgot something!

**Comment by SimonBiggs** [comment]
> seeking a way to ensure we can efficiently test Numba so that when a new release is posted, it will work for everyone. We have some ideas and will try some process changes this year.

An amazing approach to this is to deploy dev releases on PyPI. Get the whole community to make sure things work for you :slightly_smiling_face:

**Comment by SimonBiggs** [comment]
Unfortunately  the build errored like so: :slightly_frowning_face: 

```
[ 98%] Building CXX object utils/benchmark/src/CMakeFiles/benchmark.dir/benchmark_register.cc.o
In file included from /home/simon/git/llvm-11.1.0.src/utils/benchmark/src/benchmark_register.cc:15:
/home/simon/git/llvm-11.1.0.src/utils/benchmark/src/benchmark_register.h: In function ‘void AddRange(std::vector<T>*, T, T, int)’:
/home/simon/git/llvm-11.1.0.src/utils/benchmark/src/benchmark_register.h:17:30: error: ‘numeric_limits’ is not a member of ‘std’
   17 |   static const T kmax = std::numeric_limits<T>::max();
      |                              ^~~~~~~~~~~~~~
/home/simon/git/llvm-11.1.0.src/utils/benchmark/src/benchmark_register.h:17:46: error: expected primary-expression before ‘>’ token
   17 |   static const T kmax = std::numeric_limits<T>::max();
      |                                              ^
/home/simon/git/llvm-11.1.0.src/utils/benchmark/src/benchmark_register.h:17:49: error: ‘::max’ has not been declared; did you mean ‘std::max’?
   17 |   static const T kmax = std::numeric_limits<T>::max();
      |                                                 ^~~
      |                                                 std::max
In file included from /usr/include/c++/11/algorithm:62,
                 from /home/simon/git/llvm-11.1.0.src/utils/benchmark/include/benchmark/benchmark.h:175,
                 from /home/simon/git/llvm-11.1.0.src/utils/benchmark/src/internal_macros.h:4,
                 from /home/simon/git/llvm-11.1.0.src/utils/benchmark/src/check.h:8,
                 from /home/simon/git/llvm-11.1.0.src/utils/benchmark/src/benchmark_register.h:6,
                 from /home/simon/git/llvm-11.1.0.src/utils/benchmark/src/benchmark_register.cc:15:
/usr/include/c++/11/bits/stl_algo.h:3467:5: note: ‘std::max’ declared here
 3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
      |     ^~~
make[2]: *** [utils/benchmark/src/CMakeFiles/benchmark.dir/build.make:90: utils/benchmark/src/CMakeFiles/benchmark.dir/benchmark_register.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:120910: utils/benchmark/src/CMakeFiles/benchmark.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
+ exit 2
```

**Comment by jakirkham** [comment]
It might be easier to start from something like the conda-forge recipes when [building locally]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#testing-changes-locally ). There's even a PR for `llvmlite` to add Python 3.11 ( https://github.com/conda-forge/llvmlite-feedstock/pull/66 ). Though it will require more [recent source code for `llvmlite`]( https://github.com/conda-forge/llvmlite-feedstock/blob/5e11d450036a14012cdcd9accfaa9803a6cc93a7/recipe/meta.yaml#L9-L12 ), which is [easy enough to add]( https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#source-from-git )

**Comment by esc** [comment]
@SimonBiggs @sirfz @jakirkham thank you for adding to this issue! Yes, it is true that Numba/llvmlite are tricky to build(!), since this thread is about Python 3.11 support, if you are still (interested in)/(trying to) create a dev release I would encourage you to open a new issue titled "difficulties creating dev build", so that we can keep the signal-to noise ratio low for all parties subscribed to this issue. We will be happy to assist and help debug, but please the read the following comments below first.

* Regarding PyPi dev releases: this is something we are actively considering, including building wheels using GitHubActions.
* Regarding dev releases in general: we have crated a tool "numba-hatchery" which is a docker based collection of scripts to build the full llvmdev/llvmlite/Numba stack with more ease, an announcement was posted here: https://numba.discourse.group/t/ann-docker-image-for-early-testing-numba-on-python-3-11/1698
* Regarding the patches: I took another look at the documentation, and it seems it is slightly out of date. The patches that need to be applied to LLVM are always listed in the conda-recipe: https://github.com/numba/llvmlite/blob/v0.39.1/conda-recipes/llvmdev/meta.yaml#L16-L24 -- this means you don't need to apply the `llvm-lto-static.patch` and for the `llvm_11_consecutive_registers.patch` it needs to be applied using `-p2` not `-p1` (conda-build is smart enough to auto-detect the patch level, so this went unnoticed). These are bugs in the llvmlite documentation, but since we are now moving to LLVM14 they will need to be upgraded to---whatever that LLVM version needs---as part of llvmlite `v0.40.0` anyway.
* The issue encountered at https://github.com/numba/numba/issues/8304#issuecomment-1461186773 -- is due to lack of support for Numpy 1.24 -- you'll need to include the patches from the "Numpy 1.24 support" development-branch if you wish to test Numba with this.
* Lastly, due to the increasing number of supply chain attacks, we frown upon third parties posting links to binary artifacts for download like in: https://github.com/numba/numba/issues/8304#issuecomment-1461186773 -- to this end @SimonBiggs please redact the link to the zip-file you posted, thank you. 


Thank you again all for adding to this issue and thank you again everyone subscribed for your continued patience, we hope to have some Numba and llvmlite release candidates soon! 🙏 🙏 🙏

**Comment by Scolymus** [comment]
hey @esc ,

so I understand this version won't be compatible with numpy>1.24, right? I know it may not be the place to ask but, can I know what's the problem with numpy >1.24 and when there will be a version for that? I never use numba, but just install a repo dependent on numba, and after having my project working in local and integrating in a bigger project I realized numba wasn't working with numpy 1.25, which is what we use in the other project... I installed 0.57.0dev0 to allow pip to install numba while having numpy > 1.24, but I got an error that I saw is related with numpy version in a different issue:
`initialization of _internal failed without raising an exception`

**Comment by esc** [comment]
@Scolymus -- from what I can tell support for 1.24 was merged here: https://github.com/numba/numba/pull/8691 --

**Comment by esc** [comment]
@Scolymus and 1.25 support is no implemented yet and being tracked here: https://github.com/numba/numba/issues/8698

**Comment by imichaelmoore** [comment]
This issue is quite convoluted at this point, could one of the maintainers chime in and let the community know specifically what they need help with in terms of testing?  We are all watching this issue but at this point it's not clear what needs to be tested to enable a release.

**Comment by esc** [comment]
> This issue is quite convoluted at this point, could one of the maintainers chime in and let the community know specifically what they need help with in terms of testing? We are all watching this issue but at this point it's not clear what needs to be tested to enable a release. (And we want to help! 😄 )

Sure, allow me to chime in! @sklam is currently still merging development branches and we are hoping to publish a release candidate next week (maybe the week after...). As soon the projects have been tagged and the build artifacts produced, we will be very grateful for everyone to test those candidates with their respective libraries and applications (and of course... to report back any encountered issues). An announcement with instructions on how to install will be posted to our discourse (numba.discourse.group) and will post a link here too, pointing everyone in the right direction.

**Comment by kaluzniacki** [comment]
The developers have my sympathies - I very much understand how using conda makes things easier - I hope to never have to install gdal without conda, and I can install independent versions of gcc with conda. Best of luck!

**Comment by sklam** [comment]
Py3.11 branch have been merged to main by https://github.com/numba/numba/pull/8822

**Comment by pjochem4** [comment]
When will the wheel file be available on PyPi?

**Comment by sklam** [comment]
We are planning to cut an RC next week. Closing

**Comment by jakirkham** [comment]
Is there an issue to track updates on the RC and releases?

**Comment by stuartarchibald** [comment]
> Is there an issue to track updates on the RC and releases?

A Release Candidate checklist for 0.57.0rc1 is now in #8841. Please do subscribe to that if you are interested in following progress. The official RC announcement will be made on the discourse forum as usual.

**Comment by hauntsaninja** [comment]
https://github.com/hauntsaninja/python_readiness
```
λ uv tool run python_readiness --python 3.11 --package numba
numba>=0.57.0     # has_classifier_and_explicit_wheel
```
